ocaml/xapi/xapi_pool.ml

```diff
@@ -781,15 +781,37 @@ let eject ~__context ~host =
 
 	if Pool_role.is_master () then raise Cannot_eject_master
 	else begin
-		(* Fail the operation if any VMs are running here (except control domains) *)
+		(* Fail the operation if any VMs are running here (except dom0) *)
 		let my_vms_with_records = Db.VM.get_records_where ~__context ~expr:(Eq(Field "resident_on", Literal (Ref.string_of host))) in
 		List.iter (fun (_, x) -> 
-			if (not x.API.vM_is_control_domain) && x.API.vM_power_state<>`Halted
+			if (not (Helpers.is_domain_zero ~__context (Db.VM.get_by_uuid ~__context ~uuid:x.API.vM_uuid))) && x.API.vM_power_state <>`Halted
 			then begin
 				error "VM uuid %s not in Halted state and resident_on this host" (x.API.vM_uuid);
 				raise (Api_errors.Server_error(Api_errors.operation_not_allowed, ["VM resident on host"]))
 			end) my_vms_with_records;
 
+		(* collect all control domains resident on me*)
+		let controlDomainsToDestroy = List.filter (fun vmRef ->
+			let x = Db.VM.get_record ~__context ~self:vmRef in
+			if not (x.API.vM_is_control_domain) then
+				false
+			else if Db.is_valid_ref __context x.API.vM_resident_on then
+				x.API.vM_resident_on = host
+			else
+				let residentHost =
+				(try
+					let filter_valid x = List.filter (fun y -> Db.is_valid_ref __context y) x in
+					let vbds = Db.VM.get_VBDs ~__context ~self:vmRef in
+					let vdis = List.map (fun v -> Db.VBD.get_VDI ~__context ~self:v) (filter_valid vbds) in
```

Comment:
I originally thought so too, however during my testing I stumbled upon null vdi refs. I do not know whether this is a real world case (my environment contained a few hacks to simulate a halted control domain); if you can guarantee it can never happen I am happy to remove the filters.